base/inference.jl

```diff
@@ -1630,33 +1583,41 @@ function typeinf_edge(linfo::LambdaInfo, atypes::ANY, sparams::SimpleVector, nee
-    return typeinf_edge(linfo, atypes, sparams, false, true, true, caller)
+function typeinf_edge(method::Method, atypes::ANY, sparams::SimpleVector, caller)
+    return typeinf_edge(method, atypes, sparams, false, true, true, caller)
 end
-function typeinf(linfo::LambdaInfo, atypes::ANY, sparams::SimpleVector, needtree::Bool=false)
-    return typeinf_edge(linfo, atypes, sparams, needtree, true, true, nothing)
+function typeinf(method::Method, atypes::ANY, sparams::SimpleVector, needtree::Bool=false)
+    return typeinf_edge(method, atypes, sparams, needtree, true, true, nothing)
 end
 # compute an inferred (optionally optimized) AST without global effects (i.e. updating the cache)
-function typeinf_uncached(linfo::LambdaInfo, atypes::ANY, sparams::ANY; optimize::Bool=true)
-    return typeinf_edge(linfo, atypes, sparams, true, optimize, false, nothing)
-end
-function typeinf_uncached(linfo::LambdaInfo, atypes::ANY, sparams::SimpleVector, optimize::Bool)
-    return typeinf_edge(linfo, atypes, sparams, true, optimize, false, nothing)
-end
-function typeinf_ext(linfo::LambdaInfo, toplevel::Bool)
-    (code, _t, _) = typeinf_edge(linfo, linfo.specTypes, svec(), true, true, true, nothing)
-    if code.inferred
-        linfo.inferred = true
-        linfo.inInference = false
-        if linfo !== code
-            linfo.code = code.code
-            linfo.slotnames = code.slotnames
-            linfo.slottypes = code.slottypes
-            linfo.slotflags = code.slotflags
-            linfo.gensymtypes = code.gensymtypes
-            linfo.rettype = code.rettype
-            linfo.pure = code.pure
+function typeinf_uncached(method::Method, atypes::ANY, sparams::ANY; optimize::Bool=true)
+    return typeinf_edge(method, atypes, sparams, true, optimize, false, nothing)
+end
+function typeinf_uncached(method::Method, atypes::ANY, sparams::SimpleVector, optimize::Bool)
+    return typeinf_edge(method, atypes, sparams, true, optimize, false, nothing)
+end
+function typeinf_ext(linfo::LambdaInfo)
+    if isdefined(linfo, :def)
+        # method lambda - infer this specialization via the method cache
+        (code, _t, _) = typeinf_edge(linfo.def, linfo.specTypes, svec(), true, true, true, linfo)
+        if code.inferred
+            linfo.inferred = true
+            linfo.inInference = false
+            if linfo !== code
+                linfo.code = code.code
+                linfo.slotnames = code.slotnames
+                linfo.slottypes = code.slottypes
+                linfo.slotflags = code.slotflags
+                linfo.gensymtypes = code.gensymtypes
+                linfo.rettype = code.rettype
+                linfo.pure = code.pure
+            end
         end
+    else
+        # toplevel lambda - infer directly
+        frame = InferenceState(linfo, linfo.specTypes, svec(), true)
+        typeinf_loop(frame)
+        @assert frame.inferred # TODO: deal with this better
```

Comment:
it would require calling a toplevel expr from inside inference, so staged functions that thought it was OK to call `eval` (it ain't)